### PR TITLE
Support CMake header files analysis

### LINKED ISSUE #2550: Implement header file db entry location algorithm
### Description

Implement header file db entry location algorithm

### TODO
- [x] implement matching algorithm 
- [x] add debug output to show how long the search takes
- [x] unit tests

### File matching rules
The proposed algorithm is as follows:

(A) If code file:
- 1) first matching [path]\[filename with extension]
No fallback if not found

(B) If header file:
- 1) first file matching [path]\[filename without extension], else
- 2) first file matching [path], else
- 3) first item

Currently all of the matching logic is in `CompilationConfigProvider`.
The search logic for (A) has already been implemented.

### Out of scope
The initial implementation will not including caching.
The time taken to execute the matching process will debug-logged to help assess the impact.

### PR DESCRIPTION
**Pull request by rita-sonarsource**
Fixes #2550

### CODE REVIEW

File: src/Integration.Vsix.UnitTests/CFamily/CMake/CompilationDatabaseRequestFactoryTests.cs [hunk 1]
```diff
@@ -57,11 +58,33 @@ public void TryGet_NoConfig_ReturnsNull()
...
+            // However, the compilation entry is a ".c" file, so we expect the code to calculate the rules based on the entry.
+            rulesConfigProvider.Verify(x=> x.GetRulesConfiguration(SonarLanguageKeys.C), Times.Once());
+        }
+
         [TestMethod]
         public void TryGet_UnrecognizedLanguage_ReturnsNull()
         {
             const string fileName = "c:\\file.txt";
-            var compilationConfigProvider = CreateCompilationProvider(fileName, new CompilationDatabaseEntry());
+
```
Comment (rita-sonarsource) [inline src/Integration.Vsix.UnitTests/CFamily/CMake/CompilationDatabaseRequestFactoryTests.cs:85]: Changed existing tests to create a valid CreateCompilationDatabaseEntry (with a file name)

File: src/CFamily/CMake/CompilationConfigProvider.cs [hunk 2]
```diff
@@ -109,5 +120,97 @@ public CompilationDatabaseEntry GetConfig(string filePath)
...
+
+            if (entry == null)
+            {
+                logger.WriteLine(Resources.NoCompilationDatabaseEntry, filePath);
+            }
+
+            return entry;
+        }
+
+        private CompilationDatabaseEntry LocateMatchingCodeEntry(string headerFilePath, IEnumerable<CompilationDatabaseEntry> compilationDatabaseEntries)
```
Comment (rita-sonarsource) [inline src/CFamily/CMake/CompilationConfigProvider.cs:141]: I can't think of a nicer way to structure this block 🙈

File: src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequestFactory.cs [hunk 3]
```diff
@@ -47,7 +47,7 @@ public IRequest TryGet(string analyzedFilePath, CFamilyAnalyzerOptions analyzerO
             }
 
             // TODO - handle user specifying the language via a command / argument #2533
-            var languageKey = CFamilyShared.FindLanguageFromExtension(analyzedFilePath);
+            var languageKey = CFamilyShared.FindLanguageFromExtension(dbEntry.File);
```
Comment (rita-sonarsource) [inline src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequestFactory.cs:50]: We need to take the extension from the code entry, not the header file. Added an explicit test for this.

File: src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs [hunk 4]
```diff
@@ -78,7 +78,7 @@ public void WriteRequest(BinaryWriter writer)
             WriteHeader(writer);
 
             // Required inputs
-            WriteSetting(writer, "File", databaseEntry.File);
+            WriteSetting(writer, "File", Context.File);
```
Comment (rita-sonarsource) [inline src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs:81]: We need to write the analyzed header file, not the code entry -- the code entry is used only for the build parameters. I added a test for this.

File: src/Integration.Vsix.UnitTests/CFamily/CMake/CompilationDatabaseRequestTests.cs [hunk 5]
```diff
@@ -118,7 +130,7 @@ public void WriteReqest_WithCommand_ExpectedSettingWritten()
         }
 
         [TestMethod]
-        public void WriteReqest_WithArguments_ExpectedSettingWritten()
+        public void WriteRequest_WithArguments_ExpectedSettingWritten()
```
Comment (rita-sonarsource) [inline src/Integration.Vsix.UnitTests/CFamily/CMake/CompilationDatabaseRequestTests.cs:133]: Cosmetics, renaming Reqest to Request

File: src/CFamily/CMake/CompilationConfigProvider.cs (hunk 2 above)
Comment (duncanp-sonar) [reply src/CFamily/CMake/CompilationConfigProvider.cs:141]: One option would be to push more logic into each "Locate" method, so they handle the loop and the "not found" logging, returning null if they didn't find anything.

That would simplify the call to something like:
```
  var match = LocateCodeEntry(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntryWithSameName(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntryUnderRoot(codeFile, compilationDatabaseEntries) ??
    LocateFirstCodeEntry(compilationDatabaseEntries);
```

You'd probably also want to pass `codeFilesWithSameNameAndSamePath` into the "Locate" methods so they didn't have to recalculate it each time.

I'm not sure it's a big improvement, though.

**Review by duncanp-sonar — APPROVED** [review]

**Pull request merged**